### DISCUSSION TITLE
need help understanding dumpers

### DISCUSSION THREAD

**Discussion by j-carson** [psycopg/psycopg#234] [Q&A] [answered]
I have sensor data that I'm loading to a database. The sensor dumps on 32-bit word boundaries, so the API datatypes are always np.uint32 and np.float32. Depending on the particular field, the database destination oid for those uint32's may be INT or SMALLINT.

I'm trying to figure out how to create a binary dumpers for copy. I see the upgrade() method which seems to be involved, but I'm not sure how this all hooks together properly. I apparently can't register separate binary dumpers for the same source python type but different destination oid - the API doesn't seem to ask for both source and destination types. How does my "upgrade" method get told when to dump two versus four bytes?

**Comment by dvarrazzo** [comment, marked as answer]
Hello,

psycopg doesn't get information from the database about the destination column; the only information available is the Python type.

PostgreSQL can insert a value of smaller type into a column of larger type without requiring a cast. So, for instance, passing a SMALLINT to an INT field works without adding a cast. In the opposite direction, instead, a cast is required.

Psycopg uses this property for its advantage in order to dump Python ints (which can be of arbitrary size) to PostgreSQL integer fields of any type (int2, int4, int8, numeric for fields larger than the int8 range). Psycopg always specifies the smallest Postgres type that can contain the Python int that is being dumped.

As you have found, the upgrade() mechanism allows that. Some documentation is available in the [`Dumper.get_key()` API docs](https://www.psycopg.org/psycopg3/docs/api/abc.html#psycopg.abc.Dumper.get_key). Some code close to what you need is available in the [`IntDumper`](https://github.com/psycopg/psycopg/blob/7c7fe2af3542fc9537ab0ed6eace6faac15402f9/psycopg/psycopg/types/numeric.py#L126-L151) and [`IntBinaryDumper`](https://github.com/psycopg/psycopg/blob/7c7fe2af3542fc9537ab0ed6eace6faac15402f9/psycopg/psycopg/types/numeric.py#L199-L206) classes.

You should have something like (untested):

```python
class NpUint32ToInt2BinaryDumper(Dumper):
    """Convert np.uint32 to PostgreSQL int2."""
    format = Format.BINARY
    def dump(self, obj: np.uint32) -> bytes:
        # assume that the value will fit into the Postgres smallint range
        # Convert uint32 to 2 bytes in network order.
        ...

class NpUint32ToInt4BinaryDumper(Dumper):
    """Convert np.uint32 to PostgreSQL int4."""
    format = Format.BINARY
    def dump(self, obj: np.uint32) -> bytes:
        # assume that the value will fit into the Postgres int range
        # Convert uint32 to 4 bytes in network order.
        ...

class NpUint32BinaryDumper(Dumper):
    """Convert np.uint32 to PostgreSQL int2 or int4 according to size."""
    format = Format.BINARY

    _int2_key = (np.uint32, np.int2)
    _int4_key = (np.uint32, np.int4)
    _int2_dumper = NpUint32ToInt2BinaryDumper(np.uint32)
    _int4_dumper = NpUint32ToInt4BinaryDumper(np.uint32)

    def get_key(self, obj: np.uint32, format: PyFormat) -> Tuple[type, ...]:
        if obj < 2**15:
            return self._int2_key
        else:
            return self._int4_key

    def upgrade(self, obj: np.uint32, format: PyFormat) -> Dumper:
        if obj < 2**15:
            return self._int2_dumper
        else:
            return self._int4_dumper
```

**Reply by j-carson** [reply]
I get that "psycopg" doesn't know the destination column type -- but my application knows the column types. The docs for using copy in binary form say that the data must be dumped in the exact format, not just a compatible one, so I want to tell psycopg to use the right dumper for each column, but I'm confused as to how to do that.

Is there a code sample for using copy with format binary I could follow?

**Reply by dvarrazzo** [reply]
I can prepare a code sample in the next week. However, the main points are:

- create a [`Dumper`](https://www.psycopg.org/psycopg3/docs/api/adapt.html#psycopg.adapt.Dumper) subclass with the oid attribute of the target type (e.g. a dumper whose `dump()` can dump an uint16 and with oid class attribute set to `psycopg.adapters.types["int2"].oid`)
- register the dumper in the context you need (global, connection...) using `CTX.adapters.register_dumper()`; See [adapt docs](https://www.psycopg.org/psycopg3/docs/advanced/adapt.html)) for examples
- call [`Copy.set_types()`](https://www.psycopg.org/psycopg3/docs/api/cursors.html#psycopg.Copy.set_types) with the list of type names or oids (e.g.`["int2", "int2", "int4"]`)

Is this enough to put together a solution?

**Reply by j-carson** [reply]
Thank you! The set_types is the part of the puzzle I wasn't seeing.

One more question, if you don't mind: When should you use copy.write_row(...) versus copy.write(buffer)?

**Reply by dvarrazzo** [reply]
`write_row()` is when you have records of python data, such as a sequence of tuples, and you need to adapt them to postgres, like for normal queries.

`write()` is for when you have data already formatted in a way that postgres understands (for instance data coming from a previous COPY ... TO FILE operation) and you just need to load it.

**Reply by j-carson** [reply]
Thank you again!

**Comment by fabiog1901** [comment]
Hello Daniele, 

based off this conversation I was able to properly code my own Dumper class, which covers the cases when a `list` is actually a JSON object which happens to be a list and maps to datatype `JSONB`, compare to a regular list which maps to an `ARRAY`.

The solution below seems to work, and I report it here for reference but also to ask if I'm not over-engineering this? The use case at hand seems common to me, yet it requires a bespoke solution.

```python
import psycopg
from psycopg.types.array import ListDumper
from psycopg.types.json import Jsonb, JsonbDumper
from psycopg.abc import Dumper
from psycopg.pq import Format


"""
CREATE TABLE t (
    s_arr STRING[] NULL,
    j JSONB NULL
);
"""

DB_DSN = "postgres://cockroach:cockroach@localhost:26257/defaultdb?sslmode=require"


class DictJsonbDumper(JsonbDumper):
    def dump(self, obj):
        return super().dump(Jsonb(obj))


class DDumper(Dumper):
    """
    Returns the correct Dumper based on the object topology:

    list[str] or empty list -> ListDumper
    list[dict] -> DictJsonbDumper
    """

    format = Format.BINARY
    oid = None

    _dict_dumper = DictJsonbDumper(list)
    _list_dumper = ListDumper(list)

    def upgrade(self, obj, format: Format) -> Dumper:
        if obj and isinstance(obj[0], dict):
            return self._dict_dumper
        else:
            return self._list_dumper


with psycopg.connect(DB_DSN, autocommit=True) as conn:
    conn.adapters.register_dumper(set, ListDumper)
    conn.adapters.register_dumper(list, DDumper)
    conn.adapters.register_dumper(dict, DictJsonbDumper)

    with conn.cursor() as cur:
        try:

            # s_arr - uses ListDumper
            myset: set[str] = set(["a", "b", "c"])
            cur.execute("INSERT INTO t (s_arr) VALUES (%s)", (myset,))
            print("1")

            # s_arr -- uses DDumper -> ListDumper
            mylist: list[str] = ["a", "b", "c"]
            cur.execute("INSERT INTO t (s_arr) VALUES (%s)", (mylist,))
            print("2")

            # j -- uses DictJsonbDumper
            mydict: dict[str, str] = {"k": "v", "k1": "v1"}
            cur.execute("INSERT INTO t (j) VALUES (%s)", (mydict,))
            print("3")

            # s_arr -- uses DDumper -> DictJsonbDumper
            mydictlist: list[dict] = [{"k": "v", "k1": "v1"}]
            cur.execute("INSERT INTO t (j) VALUES (%s)", (mydictlist,))
            print("4")

            # s_arr -- uses DDumper -> ListDumper
            myemptylist: list[dict] = []
            cur.execute("INSERT INTO t (s_arr) VALUES (%s)", (myemptylist,))
            print("5")

            # j -- uses DDumper -> ListDumper
            myemptylist: list[dict] = []
            cur.execute("INSERT INTO t (j) VALUES (%s)", (myemptylist,))
            print("6")

        except Exception as e:
            print("\t>>>", e)
```
 
The resulting table:

```sql
root@localhost:26257/defaultdb> select * from t;                                                                                                                                                                                                              
   s_arr  |            j
----------+---------------------------
  {b,c,a} | NULL
  {a,b,c} | NULL
  NULL    | {"k": "v", "k1": "v1"}
  NULL    | [{"k": "v", "k1": "v1"}]
  {}      | NULL
  NULL    | {}
(6 rows)
```
